### DISCUSSION TITLE
PyAnatomist all the steps to run my own .py script using anatomist.api?

### DISCUSSION THREAD

**Discussion by ece7048** [brainvisa/brainvisa.github.io#121] [Q&A] [answered]
Dear Brainvisa community,

I want to install the development singularity image of brainvisa and run one python script to use anatomist API and soma API.
I am trying to do this in a university cluster with ssh.

The university cluster OS is:

NAME="Scientific Linux"
VERSION="7.9 (Nitrogen)"
ID="scientific"
ID_LIKE="rhel centos fedora"
VERSION_ID="7.7"
PRETTY_NAME="Scientific Linux 7.9 (Nitrogen)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:scientificlinux:scientificlinux:7.9:GA"
HOME_URL="http://www.scientificlinux.org//"
BUG_REPORT_URL="mailto:scientific-linux-devel@listserv.fnal.gov"
REDHAT_BUGZILLA_PRODUCT="Scientific Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.9
REDHAT_SUPPORT_PRODUCT="Scientific Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.9"

I am using python==3.8.3
and I am trying to install in an anaconda3 environment.

Till now I did the follow steps:

- download the casa-dev-5.0-11.sif and casa-dev-5.0-11.sif.json
- singularity run -c -B casa_distro/brainvisa-opensource-master:/casa/setup ./casa_distro/casa-dev-5.0-11.sif branch=master distro=bvdev
- export PATH="$HOME/casa_distro/brainvisa-opensource-master/bin:$PATH"
- bv_maker (p.s. the bv_maker doc is not implemented correctly but the 3 previous step was ok!!)
- pip install pyqt5

All these steps are successfully implemented.

When I am trying to run:

python3
>> import anatomist.api as anatomist

**Traceback (most recent call last):
File "", line 1, in
ModuleNotFoundError: No module named 'anatomist'**

if i try:

ipython --gui=qt
>> import anatomist.api as anatomist
**ModuleNotFoundError Traceback (most recent call last)
in----> 1 import anatomist.api as anatomist
ModuleNotFoundError: No module named 'anatomist'**

Any idea why there is no bidding with the PyAnatomist API??

Thank you a lot!

**Comment by c-langlet** [comment]
Hi, try to enter the brainvisa container first:
```
bv bash
python3
...
```

**Comment by denisri** [comment, marked as answer]
Exactly: you have built things inside the container, but try to run them outside of the container. As @c-langlet said, you have to start either bash, or python, or ipython, inside the container:
```
bv bash
```
or:
```
bv python3
```
or:
```
bv ipython3
```

**Comment by ece7048** [comment]
Thank you both!! 
This work!
I need to use **pip install .** and a setup.py file  to install my code.
Can I do pip install inside the bv bash?

When I tried this it prints a huge error:
    Running setup.py install for mysql-connector ... error
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python2 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-yzYVvk/mysql-connector/setup.py'"'"'; __file__='"'"'/tmp/pip-install-yzYVvk/mysql-connector/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-b9DVW9/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/include/python2.7/mysql-connector
         cwd: /tmp/pip-install-yzYVvk/mysql-connector/
    Complete output (88 lines):
    running install
    Not Installing C Extension
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-2.7
    creating build/lib.linux-x86_64-2.7/mysql
    copying lib/mysql/__init__.py -> build/lib.linux-x86_64-2.7/mysql
    creating build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/version.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/utils.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/protocol.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/pooling.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/optionfiles.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/network.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/errors.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/errorcode.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/dbapi.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/custom_types.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/cursor_cext.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/cursor.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/conversion.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/constants.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/connection_cext.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/connection.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/charsets.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/catch23.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/authentication.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/abstracts.py -> build/lib.linux-x86_64-2.7/mysql/connector
    copying lib/mysql/connector/__init__.py -> build/lib.linux-x86_64-2.7/mysql/connector
    creating build/lib.linux-x86_64-2.7/mysql/connector/locales
    copying lib/mysql/connector/locales/__init__.py -> build/lib.linux-x86_64-2.7/mysql/connector/locales
    creating build/lib.linux-x86_64-2.7/mysql/connector/locales/eng
    copying lib/mysql/connector/locales/eng/client_error.py -> build/lib.linux-x86_64-2.7/mysql/connector/locales/eng
    copying lib/mysql/connector/locales/eng/__init__.py -> build/lib.linux-x86_64-2.7/mysql/connector/locales/eng
    creating build/lib.linux-x86_64-2.7/mysql/connector/django
    copying lib/mysql/connector/django/validation.py -> build/lib.linux-x86_64-2.7/mysql/connector/django
    copying lib/mysql/connector/django/schema.py -> build/lib.linux-x86_64-2.7/mysql/connector/django
    copying lib/mysql/connector/django/operations.py -> build/lib.linux-x86_64-2.7/mysql/connector/django
    copying lib/mysql/connector/django/introspection.py -> build/lib.linux-x86_64-2.7/mysql/connector/django
    copying lib/mysql/connector/django/features.py -> build/lib.linux-x86_64-2.7/mysql/connector/django
    copying lib/mysql/connector/django/creation.py -> build/lib.linux-x86_64-2.7/mysql/connector/django
    copying lib/mysql/connector/django/compiler.py -> build/lib.linux-x86_64-2.7/mysql/connector/django
    copying lib/mysql/connector/django/client.py -> build/lib.linux-x86_64-2.7/mysql/connector/django
    copying lib/mysql/connector/django/base.py -> build/lib.linux-x86_64-2.7/mysql/connector/django
    copying lib/mysql/connector/django/__init__.py -> build/lib.linux-x86_64-2.7/mysql/connector/django
    creating build/lib.linux-x86_64-2.7/mysql/connector/fabric
    copying lib/mysql/connector/fabric/connection.py -> build/lib.linux-x86_64-2.7/mysql/connector/fabric
    copying lib/mysql/connector/fabric/caching.py -> build/lib.linux-x86_64-2.7/mysql/connector/fabric
    copying lib/mysql/connector/fabric/balancing.py -> build/lib.linux-x86_64-2.7/mysql/connector/fabric
    copying lib/mysql/connector/fabric/__init__.py -> build/lib.linux-x86_64-2.7/mysql/connector/fabric
    creating build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/statement.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/result.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/protocol.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/expr_unparser.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/expr.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/errors.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/errorcode.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/dbdoc.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/crud.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/constants.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/connection.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/compat.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/charsets.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/authentication.py -> build/lib.linux-x86_64-2.7/mysqlx
    copying lib/mysqlx/__init__.py -> build/lib.linux-x86_64-2.7/mysqlx
    creating build/lib.linux-x86_64-2.7/mysqlx/locales
    copying lib/mysqlx/locales/__init__.py -> build/lib.linux-x86_64-2.7/mysqlx/locales
    creating build/lib.linux-x86_64-2.7/mysqlx/locales/eng
    copying lib/mysqlx/locales/eng/client_error.py -> build/lib.linux-x86_64-2.7/mysqlx/locales/eng
    copying lib/mysqlx/locales/eng/__init__.py -> build/lib.linux-x86_64-2.7/mysqlx/locales/eng
    creating build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/t.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/mysqlx_sql_pb2.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/mysqlx_session_pb2.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/mysqlx_resultset_pb2.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/mysqlx_pb2.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/mysqlx_notice_pb2.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/mysqlx_expr_pb2.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/mysqlx_expect_pb2.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/mysqlx_datatypes_pb2.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/mysqlx_crud_pb2.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/mysqlx_connection_pb2.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    copying lib/mysqlx/protobuf/__init__.py -> build/lib.linux-x86_64-2.7/mysqlx/protobuf
    running install_lib
    creating /usr/local/lib/python2.7/dist-packages/mysqlx
    error: could not create '/usr/local/lib/python2.7/dist-packages/mysqlx': Function not implemented
    ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/bin/python2 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-yzYVvk/mysql-connector/setup.py'"'"'; __file__='"'"'/tmp/pip-install-yzYVvk/mysql-connector/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-b9DVW9/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/include/python2.7/mysql-connector Check the logs for full command output.

Thank you :)

**Comment by denisri** [comment]
Uh oh, several things:
- you probably want to use python3, not python2. `casa-dev-5.0` images contain both but use python2 by default. To use python3, either:
  - edit (inside the container) `/casa/host/conf/bv_maker.cfg` and add, in the `[ build ]` section, the followng line:
    ```
    cmake_options += -DPYTHON_EXECUTABLE=python3
    ```
  - or use `casa-dev-5.3.sif` images, which are based on Ubuntu 22.04 and use python3 by default. I would recommend this solution actually.
 
  You will have to rebuild everything, sorry. You'd better erase everything in `/casa/host/build` to make sure there will be no cmake cache conflicts, and run `bv_maker configure build` again.
  `casa-dev-5.0.sif` images are based on Ubuntu 18.04 and contain python 3.6.
  `casa-dev-5.3.sif` images are based on Ubuntu 22.04 and contain python 3.10.

- don't do `pip install pyqt5` as it will install incompatible Qt libraries. It's already installed in the image. However I guess you only have done it outside of the container before we told you to run things inside it.
- You can use pip inside the container, but make sure to use `pip3` (or `python3 -m pip`) especially in 5.0 images otherwise it will be pip for python2, as it appears in the log messages you have copied above.
- You need to do a **user install** in pip. I think pip3 does it by default if it cannot write to `/usr/local`, but this was not working using pip2 as in your try above. If needed use the `--user` option in pip.

**Reply by ece7048** [reply]
Thank you a lot!! This is very helpful!
Unfortunately. When I tried the casa-dev-5.3-10 and 5.3-12 in the bv_maker build there is always this error:

_[ 15%] Built target aims-free_target_30
[ 18%] Built target axon_target_3
[ 21%] Built target anatomist-free_target_13
[ 26%] Built target axon_target_10
[ 26%] Built target axon_target_4
[ 31%] Built target brainvisa-spm_target_1
[ 73%] Built target brainvisa-share_target_2
make: *** [Makefile:166: all] Error 2
Traceback (most recent call last):
  File "/casa/host/build/python/brainvisa/maker/commands.py", line 71, in process
    logs = getattr(o, method)(*meth_args,
  File "/casa/host/build/python/brainvisa/maker/build.py", line 665, in build
    system(cwd=self.directory, *(['make'] + self.make_options),
  File "/casa/host/build/python/brainvisa/maker/subprocess.py", line 59, in system
    six.reraise(*error)
  File "/usr/lib/python3/dist-packages/six.py", line 718, in reraise
    raise value.with_traceback(tb)
  File "/casa/host/build/python/brainvisa/maker/subprocess.py", line 48, in system
    raise OSError()
OSError: Command failed: 'make' '-j32'
From directory: /casa/host/build

bv_maker summary:
FAILED      step build: /casa/host/build, started: 2023/01/11 16:21, stopped: 2023/01/11 16:22
There were errors._

I did the bv_maker after bv bash and without bv_bash. The same result.

Any help Please!!

Thank you,
 Mike

**Comment by denisri** [comment]
The actual error is earlier, we don't see it here in your log...

**Reply by ece7048** [reply]
/usr/lib/qt5/bin/uic: error while loading shared libraries: libQt5Core.so.5: cannot open shared object file: No such file or directory
make[2]: *** [build_files/anatomist-free/src/lib/CMakeFiles/run_uic.dir/build.make:80: include/anatomist/ui/ui_color_rendering.h] Error 127

I think this is the main error.

Thank you :)

**Comment by denisri** [comment]
We recently ran into a very similar problem in the computing center of our institution. This computing center imposes to use a different type of container (PCOCC), not singularity, so we had to convert using several intermediate steps. Thus we were thinking the problem was due to a container image conversion problem, or to that container system, but as you experience the same problem, it might be more general. Moreover the PCOCC maintainer has just suggested us that it might be the host system kernel (which is shared with the container) which is too old, and that Qt libraries may check a compatibility version, and fail this test.
The suggested solution would be to remove the ABI version requirements in Qt libs, and rebuild images. We have not tried it yet, but will do so in the next days. If that works, we will rebuild new images 5.3. It may or may not fix the problem depending on the features actually available in the host system kernel.
Which version of the linux kernel is your host system using ? (run `uname -r` to know it)

**Reply by ece7048** [reply]
Thank you.
This is the output:
3.10.0-1160.80.1.el7.x86_64

**Comment by denisri** [comment]
see https://github.com/brainvisa/casa-distro/issues/324

**Comment by denisri** [comment]
I have rebuilt new images with the ABI tag stripped (in principle), but I could not test them so far because I cannot make the container system work in the computing center any longer at the moment. We'll test next week, but if you like to try them also, you can.
You can pull updated images using `casa_distro pull_image`.

**Reply by denisri** [reply]
The new images seem to work for us. Now it probably actually depends on your kernel, but you seem to be using RHEL7, same as in the computing center we are using, so it might behave the same. And I have patched Qt libraries, but I don't know if other system libraries could be affected by the same problem.

**Reply by ece7048** [reply]
Thank you for all the responses and answers. 
The pip3 install works for the casa_dev5.0.-11 based on your 1st suggestion  to rebuild using python3. The casa-dev-5.3.10 did not build correctly.

**Reply by denisri** [reply]
I was taking of [casa-dev-5.3-13.sif](https://brainvisa.info/download/casa-dev-5.3-13.sif) (rebuilt last friday), not casa-dev-5.3.10 which is older and does not include the modifications. However it should work using casa-dev-5.0 or 5.1 which are based on an older system (Ubuntu 18.04). Just I won't know if my fix works (for you and possibly others).

**Reply by ece7048** [reply]
The only new issue is that when I try to run the python script code of anatomist.api return the above error:

QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-mm2703'
qt.qpa.screen: QXcbConnection: Could not connect to display 
Could not connect to any X display.

**Reply by denisri** [reply]
It seems that the container cannot connect to a X server (graphics server). Either it is run in non-graphical mode (unlikely with singularity, it is done automatically by default), or you are working on a remote machine, typically using a ssh connection, without display options (ssh -X or ssh -Y). It should be the same outside of the container using any graphical command ?
